src/lib/form-field/form-field.ts

```diff
@@ -0,0 +1,289 @@
+/**
+ * @license
+ * Copyright Google Inc. All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import {
+  AfterContentChecked,
+  AfterContentInit,
+  AfterViewInit,
+  ChangeDetectionStrategy,
+  ChangeDetectorRef,
+  Component,
+  ContentChild,
+  ContentChildren,
+  ElementRef,
+  Inject,
+  Input,
+  Optional,
+  QueryList,
+  ViewChild,
+  ViewEncapsulation,
+} from '@angular/core';
+import {animate, state, style, transition, trigger} from '@angular/animations';
+import {coerceBooleanProperty} from '../core';
+import {
+  getMdFormFieldDuplicatedHintError,
+  getMdFormFieldMissingControlError,
+  getMdFormFieldPlaceholderConflictError,
+} from './form-field-errors';
+import {
+  FloatPlaceholderType,
+  MD_PLACEHOLDER_GLOBAL_OPTIONS,
+  PlaceholderOptions
+} from '../core/placeholder/placeholder-options';
+import {startWith} from '@angular/cdk/rxjs';
+import {MdError} from './error';
+import {MdFormFieldControl} from './form-field-control';
+import {MdHint} from './hint';
+import {MdPlaceholder} from './placeholder';
+import {MdPrefix} from './prefix';
+import {MdSuffix} from './suffix';
+
+
+let nextUniqueId = 0;
+
+
+/** Container for form controls that applies Material Design styling and behavior. */
+@Component({
+  moduleId: module.id,
+  // TODO(mmalerba): the input-container selectors and classes are deprecated and will be removed.
+  selector: 'md-input-container, mat-input-container, md-form-field, mat-form-field',
+  templateUrl: 'form-field.html',
+  // MdInput is a directive and can't have styles, so we need to include its styles here.
+  // The MdInput styles are fairly minimal so it shouldn't be a big deal for people who aren't using
+  // MdInput.
+  styleUrls: ['form-field.css', '../input/input.css'],
+  animations: [
+    trigger('transitionMessages', [
```

Comment:
I know that this is just a refactoring, but maybe we should consider getting the placeholder animation to go through Angular as well? This allows users to disable it a bit easier.